README.md
# Trinidad Database Pool

A bunch of Trinidad extensions to support database connection pooling on top of
the underlying Apache Tomcat container (using Tomcat's JDBC Connection Pool).

Please note, that such pools are usually configured as "global" and thus
shareable by multiple applications deployed on to of the Trinidad server.

Also most Java issues caused by JDBC drivers packaged with the application (e.g.
`java.lang.UnsatisfiedLinkError` with SQLite3's JDBC driver) that JRuby inherits
should go away when database connections are managed externally.

http://tomcat.apache.org/tomcat-7.0-doc/jdbc-pool.html

## Available Pools

* MySQL (trinidad_mysql_dbpool_extension)
* PostgreSQL (trinidad_postgresql_dbpool_extension)
* SQLite (trinidad_sqlite_dbpool_extension)
* MS-SQL (trinidad_mssql_dbpool_extension) using (unofficial) jTDS driver
* Generic (trinidad_generic_dbpool_extension) in case your DB ain't supported

## Usage

* Install the gem e.g. `jruby -S gem install trinidad_mysql_dbpool_extension`
* Configure the pool with Trinidad's configuration file e.g. :

```yml
  web_apps:
    default:
      extensions:
        mysql_dbpool:                   # EXTENSION NAME AS KEY
          jndi: 'jdbc/MySampleDB'       # name (linked with database.yml)
          url: 'localhost:3306/sample'  # database URL (or full jdbc: URL)
          username: root
          password: root
          maxActive:  100  # maximum number of connections managed by the pool
          initialSize: 10  # initial number of connections created in the pool
          maxWait:  10000  # ms the pool waits for a connection to be returned
```

**jndi**, **url** are mandatory (as well **username** and **password** if your
database server + driver requires authentication), while other supported
configuration options might be found here in Tomcat's JDBC pool documentation :

http://tomcat.apache.org/tomcat-7.0-doc/jdbc-pool.html#Common_Attributes

**NOTE:** starting version 0.7 we switched to using Tomcat's JDBC pool which
is simpler and more performant pool designed specifically for Tomcat/Trinidad.

If you insist on using [DBCP](http://commons.apache.org/dbcp/configuration.html)
set the (deprecated) *pool: dbcp* option along side your pool configuration ...

If you happen to be using *SQLite* on production you'll need an absolute path :

```yml
  extensions:
    sqlite_dbpool:
      jndi: jdbc/FileDB
      url: <%= File.expand_path('db/production.db') %>
```

* Configure the (Rails) application to use JNDI with *config/database.yml*

```yml
production:
  adapter: mysql
  jndi: java:/comp/env/jdbc/MySampleDB
  # (Java) JDBC driver class name is detected for built-in adapters that
  # activerecord-jdbc-adapter supports, specify for a "generic" adapter
  # driver: com.mysql.jdbc.Driver
```

do not forget to delete **pool** setup part since there's no need for it ...

## Generic Pool

If there's no "official" pool for your database, or would like to use a
different version of a JDBC driver for a supported Trinidad pool, this allows
you to completely customize a pool.

NOTE: You will need a JDBC driver for your database, check the driver DB if
unsure about how to get one http://developers.sun.com/product/jdbc/drivers

Sample configuration for a DB2 database :

```yml
---
  extensions:
    generic_dbpool:                          # EXTENSION NAME AS KEY
      jndi: 'jdbc/MyDB'                      # JNDI name
      url: 'jdbc:db2://127.0.0.1:50000/MYDB' # specify full jdbc: URL
      username: 'mydb'                       # database username
      password: 'pass'                       # database password
      driverPath: '/opt/IBM/DB2/db2jcc4.jar' # leave out if on class-path
      driverName: com.ibm.db2.jcc.DB2Driver  # resolved from driverPath jar
```

Beyond standard configuration options there's 2 specific options here :

* **driverPath** should be a path to your JDBC driver archive, might leave that
  out but make sure it's on the (shared) class-path for each and every deployed
  application that requires it. You might specify multiple jars using the
  `Dir.glob` syntax or by separating paths using the `:` separator.

  Also in case *driverPath* is omitted you'll need to specify a *driverName* !

* **driverName** the class name implementing the JDBC driver interface, if
  you're not sure check the .jar for a META-INF/services/java.sql.Driver file.
  If present that file contains the class-name and you might leave *driverName*
  blank if you specified the jar path using *driverPath* but you're going still
  going to need the driver name in your *database.yml*

## Copyright

Copyright (c) 2013 [Team Trinidad](https://github.com/trinidad).
See LICENSE (http://en.wikipedia.org/wiki/MIT_License) for details.


History.txt
== 0.7.0 (2013-12-11)

**NOTE:** tomcat-dbcp.jar contains patched Commons-DBCP pool implementation !
We're now switching to tomcat-jdbc.jar (which we kind of assumed we're using).

 * packaged in tomcat-jdbc.jar (from 7.0.47) bundle version 1.1.0.1
 * updated tomcat-dbcp.jar (keeping it for a while for compatibility)

== 0.5.0 (2013-06-28)

 * updated tomcat-dbcp.jar to 7.0.41 ... due recent fixes in 7.0.39 / 7.0.40 :
  - major statement leak https://issues.apache.org/bugzilla/show_bug.cgi?id=54732
 * no longer bundling driver jars instead depending on jdbc-xxx gems :
  - trinidad_mysql_dbpool_extension -> jdbc-mysql
  - trinidad_postgresql_dbpool_extension - jdbc-postgres
  - trinidad_sqlite_dbpool_extension - jdbc-sqlite3
  - trinidad_mssql_dbpool_extension - jdbc-jtds

== 0.4.2 (2012-12-22)

 * updated tomcat-dbcp.jar to 7.0.32
 * support multiple jar files (and globbing) with driver path
 * trinidad_mssql_dbpool_extension jTDS jar updated to 1.3.0

== 0.4.1 (2012-11-29)

* updated driver jars for following :
  - trinidad_mysql_dbpool_extension
  - trinidad_postgresql_dbpool_extension
  - trinidad_mssql_dbpool_extension
* added trinidad_sqlite_dbpool_extension

== 0.4.0 (2012-05-31)

* revised trinidad_generic_dbpool_extension :
  - driver configuration option aliased as driverName
  - new driverPath option for specifying driver jar in configuration
  - allow driverName to be auto resolved when driverPath specified
* trinidad_postgresql_dbpool_extension with driver updated to 9.1-902
* trinidad_mysql_dbpool_extension with mysql-connector-j 5.1.20
* code cleanup, get rid of jeweler, update tomcat-dbcp.jar for trinidad_dbpool

== 0.3.0 (2011-07-01)

* Oracle connection pool extension
* Mssql connection pool extension
* Postgresql syntax error fix
* Add support for multiple pools per db extension

== 0.2.0 (2010-07-29)

* Update trinidad_dbpool_extension common gem to support Tomcat 7

== 0.1.1 (2010-05-05)

* Update trinidad_dbpool_extension common gem
* Remove adding jndi context and resource to the global tomcat context

== 0.1.0 (2010-04-04)

* First release
* General connection pool extension
* Mysql connection pool extension
* Postgresql connection pool extension


lib/trinidad_dbpool/webapp_extension.rb
require 'trinidad'

module Trinidad
  module Extensions
    class DbpoolWebAppExtension < WebAppExtension

      def configure(tomcat, context)
        case @options
        when Hash
          [create_resource(tomcat, context, @options)]
        when Array
          @options.map { |opts| create_resource tomcat, context, opts }
        end
      end

      protected

      def create_resource tomcat, context, options
        jndi, url = options.delete(:jndi), options.delete(:url)
        url = protocol + url unless %r{^#{protocol}} =~ url
        options[:url] = url

        resource_factory = options.key?(:factory) ? options.delete(:factory) :
          'org.apache.tomcat.jdbc.pool.DataSourceFactory'
        if pool = options.delete(:pool) # pool: dbcp (backwards compatibility)
          begin
            load File.expand_path("../../../trinidad-libs/tomcat-#{pool}.jar", __FILE__)
            resource_factory = nil
          rescue LoadError
            context.logger.warn "The `pool: #{pool}` option is not supported, please remove it"
          else
            context.logger.info "Using deprecated `pool: #{pool}` configuration option, consider removing it"
          end
        end

        load_driver

        driver_name = options.delete(:driver) || options.delete(:driverName) ||
                      self.driver_name

        # <Resource name="jdbc/MyDB"
        #           auth="Container"
        #           type="javax.sql.DataSource"
        #           url="jdbc:mysql://localhost:3306/mydb"
        #           driverClassName="com.mysql.jdbc.Driver"
        #           maxActive="100" maxIdle="30" maxWait="10000"
        #           username="root" password="secret" />
        resource = Trinidad::Tomcat::ContextResource.new
        resource.set_auth(options.delete(:auth)) if options.key?(:auth)
        resource.set_description(options.delete(:description)) if options.key?(:description)
        resource.set_name(jndi)
        resource.set_type(options.delete(:type) || 'javax.sql.DataSource')
        resource.set_property('factory', resource_factory) if resource_factory
        resource.set_property('driverClassName', driver_name)
        options.each { |key, value| resource.set_property(key.to_s, value.to_s) }

        context.naming_resources.add_resource(resource)
        context.naming_resources = resource.naming_resources

        resource
      end

      def load_driver; end

    end
  end
end
